### PR TITLE
[SPARK-29945][SQL] do not handle negative sign specially in the parser

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Remove the special handling of the negative sign in the parser (interval literal and type constructor)

### Why are the changes needed?

The negative sign is an operator (UnaryMinus). We don't need to handle it specially, which is kind of doing constant folding at parser side.

### Does this PR introduce any user-facing change?

The error message becomes a little different. Now it reports type mismatch for the `-` operator.

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @maropu @MaxGekk @yaooqinn

File: sql/core/src/test/resources/sql-tests/inputs/literals.sql [hunk 1]
```diff
@@ -107,7 +107,11 @@ select integer '2147483648';
 
 -- awareness of the negative/positive sign before type
 select -integer '7';
+select +integer '7';
+select +date '1999-01-01';
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/literals.sql]: We should have a followup to forbid this. I've checked pgsql and it fails.

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 2]
```diff
@@ -239,15 +239,15 @@ NULL
 -- !query 29
 select -interval '-1 month 1 day -1 second'
 -- !query 29 schema
-struct<1 months -1 days 1 seconds:interval>
+struct<(- -1 months 1 days -1 seconds):interval>
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out]: We should change the default alias of interval literal in a followup.
```
scala> sql("select interval '1 day', date '2000-10-10'").show
+------+-----------------+
|1 days|DATE '2000-10-10'|
+------+-----------------+
|1 days|       2000-10-10|
+------+-----------------+
```
We should follow date and pick `INTERVAL '1 days'`

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out [hunk 3]
```diff
@@ -239,15 +239,15 @@ NULL
...
 -- !query 29 output
 1 months -1 days 1 seconds
 
 
 -- !query 30
 select -interval -1 month 1 day -1 second
 -- !query 30 schema
-struct<1 months -1 days 1 seconds:interval>
+struct<(- -1 months 1 days -1 seconds):interval>
```
Comment (yaooqinn) [inline sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out]: the schema looks weird here

File: sql/core/src/test/resources/sql-tests/inputs/literals.sql [hunk 4]
```diff
@@ -107,7 +107,11 @@ select integer '2147483648';
...
+select +timestamp '1999-01-01';
+select +x'2379ACFe';
+-- can't negate date/timestamp/binary
 select -date '1999-01-01';
 select -timestamp '1999-01-01';
 select -x'2379ACFe';
-select +integer '7';
```
Comment (yaooqinn) [inline sql/core/src/test/resources/sql-tests/inputs/literals.sql]: Does this case not supported?

File: sql/core/src/test/resources/sql-tests/inputs/literals.sql (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/literals.sql]: do we have a typical identifier that can match `+date` as a whole?

Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/literals.sql]: no, but the parser simply ignores the `+` sign.

File: sql/core/src/test/resources/sql-tests/inputs/literals.sql (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/literals.sql]: it's supported: https://github.com/apache/spark/pull/26578/files#diff-3adf2824cec2162e632a5114d3d710cdR110

File: sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/ansi/interval.sql.out]: will be fixed by https://github.com/apache/spark/pull/26579

**Review by HyukjinKwon — APPROVED** [review]
Looks fine in general.

**Comment by HyukjinKwon** [comment]
retest this please

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.

**Pull request closed without merging**